bsp/espressif/esp/src/hals/esp32-c3/gpio.zig

```diff
@@ -0,0 +1,148 @@
+            .FUN_WPU = @intFromBool(config.pullup_enable),
+            .FUN_DRV = @intFromEnum(config.drive_strength),
+            .FILTER_EN = @intFromBool(config.input_filter_enable),
+            .FUN_IE = @intFromBool(config.input_enable),
+            .MCU_SEL = 1,
+        });
+
+        GPIO.FUNC_OUT_SEL_CFG[number].write(.{
+            .OUT_SEL = 0x80,
+            .OEN_SEL = @intFromBool(config.output_enable),
+            .INV_SEL = @intFromBool(config.output_invert),
+            .OEN_INV_SEL = 0,
+            .padding = 0,
+        });
+
+        GPIO.PIN[number].modify(.{
+            .PIN_PAD_DRIVER = @intFromBool(config.open_drain),
+        });
+
+        // NOTE: Assert that the USB_SERIAL_JTAG peripheral which uses pins GPIO18 and GPIO19
+        //       is disabled and USB pullup/down resistors are disabled.
+        if (number == 18 or number == 19) {
+            const usb_conf0 = peripherals.USB_DEVICE.CONF0.read();
+            std.debug.assert(usb_conf0.USB_PAD_ENABLE == 0 and usb_conf0.DP_PULLUP == 0 and usb_conf0.DP_PULLDOWN == 0 and usb_conf0.DM_PULLUP == 0 and usb_conf0.DM_PULLDOWN == 0);
+        }
+
+        GPIO.ENABLE_W1TS.write(.{ .ENABLE_W1TS = @as(u17, 1) << number, .padding = 0 });
+
+        return Pin{
+            .number = number,
+        };
+    }
+
+    pub fn deinit(self: Pin) void {
+        IO_MUX.GPIO[self.number].write_raw(0x00);
+        GPIO.FUNC_OUT_SEL_CFG[self.number].write_raw(0x00);
+        GPIO.PIN[self.number].write_raw(0x00);
+        GPIO.ENABLE_W1TC.write(.{ .ENABLE_W1TC = @as(u17, 1) << self.number, .padding = 0 });
+    }
+
+    pub fn set_output_enable(self: Pin, enable: bool) void {
+        GPIO.FUNC_OUT_SEL_CFG[self.number].modify(.{
+            .OEN_SEL = @intFromBool(enable),
+        });
+    }
+
+    pub fn set_open_drain(self: Pin, enable: bool) void {
+        GPIO.PIN[self.number].modify(.{
+            .PIN_PAD_DRIVER = @intFromBool(enable),
+        });
+    }
+
+    pub fn set_pullup(self: Pin, enable: bool) void {
+        IO_MUX.GPIO[self.number].modify(.{
+            .MCU_WPU = @intFromBool(enable),
+        });
+    }
+
+    pub fn set_pulldown(self: Pin, enable: bool) void {
+        IO_MUX.GPIO[self.number].modify(.{
+            .MCU_WPD = @intFromBool(enable),
+        });
+    }
+
+    pub fn set_output_invert(self: Pin, enable: bool) void {
+        GPIO.FUNC_OUT_SEL_CFG[self.number].modify(.{
+            .OEN_INV_SEL = enable,
+        });
+    }
+
+    pub fn set_output_drive_strength(self: Pin, strength: DriveStrength) void {
+        IO_MUX.GPIO[self.number].modify(.{
+            .FUN_DRV = @intFromEnum(strength),
+        });
+    }
+
+    pub fn set_input_filter(self: Pin, enable: bool) void {
+        IO_MUX.GPIO[self.number].modify(.{
+            .FILTER_EN = @intFromBool(enable),
+        });
+    }
+
+    pub fn write(self: Pin, level: Level) void {
+        std.debug.assert(GPIO.FUNC_OUT_SEL_CFG[self.number].read().OEN_SEL == 1);
```

Comment:
IMO I don't think this assert is required and could potentially hurt flexibility. Maybe someone for some reason wants to "pre-set" their output register before enabling the GPIO via `OEN_SEL`?

Reply:
I would keep this in, as forgetting to enable something before using it is so extremely common that not asserting here would probably waste too much of people's time.

Generally, yes I agree that only invariants should be asserted and all other errors should be returned for the user to handle, but I believe very few people expect `pin.write(Level.high)` to be fallible.

Reply:
Alternative for setting initial states: Make `Config` also contain optional `initial_output`, and have `set_output_enable()` also take that parameter?

Reply:
Yup agreed, I think in this case it's cool to leave in.